extensions/amp-install-serviceworker/0.1/amp-install-serviceworker.js

```diff
@@ -105,23 +122,193 @@ export class AmpInstallServiceWorker extends AMP.BaseElement {
+        'data-no-service-worker-fallback-shell-url');
+    shellUrl = removeFragment(shellUrl);
+    let urlMatchExpr;
+    try {
+      urlMatchExpr = new RegExp(urlMatch);
+    } catch (e) {
+      throw user().createError(
+          'Invalid "data-no-service-worker-fallback-url-match" expression', e);
+    }
+    user().assert(winUrl.origin == parseUrl(shellUrl).origin,
+        'Shell origin "%s" must be the same as source origin "%s"',
+        shellUrl, winUrl.href);
+
+    // Install URL rewriter.
+    this.urlRewriter_ = new UrlRewriter_(ampdoc, urlMatchExpr, shellUrl);
+
+    // Cache shell.
+    this.waitToPreloadShell_(shellUrl);
+  }
+
+  /**
+   * @param {string} shellUrl
+   * @return {!Promise}
+   * @private
+   */
+  waitToPreloadShell_(shellUrl) {
+    // Ensure that document is loaded and visible first.
+    const ampdoc = this.getAmpDoc();
+    const win = this.win;
+    return this.loadPromise(win).then(() => {
+      return viewerForDoc(ampdoc).whenFirstVisible();
+    }).then(() => {
+      this.deferMutate(() => this.preloadShell_(shellUrl));
+    });
+  }
+
+  /**
+   * @param {string} shellUrl
+   * @private
+   */
+  preloadShell_(shellUrl) {
+    const win = this.win;
+
+    // Preload the shell by via an iframe with `#preload` fragment.
+    const iframe = win.document.createElement('iframe');
+    iframe.id = 'i-amp-shell-preload';
+    iframe.setAttribute('src', shellUrl + '#preload');
+
+    // Make the iframe hidden.
+    iframe.setAttribute('hidden', 'true');
+    toggle(iframe, false);
+
+    // Restrict what this iframe can do: not much beyond precaching the
+    // resources. Unlike Chrome, Safari does not precan scripts w/o
+    // `allow-scripts` sandbox.
+    iframe.setAttribute('sandbox', 'allow-scripts allow-same-origin');
+
+    // Remove iframe once loaded.
+    const loaded = this.loadPromise(iframe);
+    loaded.then(() => {
+      removeElement(iframe);
+    });
+
+    // Start the preload.
+    this.element.appendChild(iframe);
+  }
+}
+
+
+/**
+ * URL Rewriter intercepts all navigations and, if within the parameters,
+ * rewrites the URL to go via shell.
+ */
+class UrlRewriter_ {
+
+  /**
+   * @param {!../../../src/service/ampdoc-impl.AmpDoc} ampdoc
+   * @param {!RegExp} urlMatchExpr
+   * @param {string} shellUrl
+   */
+  constructor(ampdoc, urlMatchExpr, shellUrl) {
+    /** @const {!Window} */
+    this.win = ampdoc.win;
+    /** @const @private {!RegExp} */
+    this.urlMatchExpr_ = urlMatchExpr;
+    /** @const @private {string} */
+    this.shellUrl_ = shellUrl;
+
+    /** @const @private {!Location} */
+    this.shellLoc_ = parseUrl(shellUrl);
+
+    listen(ampdoc.getRootNode(), 'click', this.handle_.bind(this));
+  }
+
+  /**
+   * @param {?Event} event
+   * @private
+   */
+  handle_(event) {
+    // Check event and target.
+    if (event.defaultPrevented) {
+      return;
+    }
+    const target = closestByTag(dev().assertElement(event.target), 'A');
```

Comment:
I thought about this. But I decided to steer away from adding more code into core. Especially that code would look very much like multiple `click` handlers work anyway. Also, the similarity between doc-click and this method are somewhat superficial.